File: src/klangk/klangkd-tests/e2e-tests/_e2e_env.py
"""Shared hermetic env helper for E2E test suites (#1526).

Every E2E suite that launches a subprocess (``runtestserver.py``, ``klangkd``,
or ``klangk``) must build the child's env from :func:`clean_env`, **not**
``{**os.environ, ...}``. A stray ``KLANGKD_*`` var in the CI runner's env
(or one leaked by a prior test) silently becomes the child's config and can
change test results — ``clean_env`` strips all config-affecting prefixes so
the child sees only what the test explicitly sets.

Strips (case-insensitive prefix match): ``KLANGK``, ``_KLANGK``,
``KLANGKC``, ``LOGFIRE``. OS-essential vars (``PATH``, ``HOME``, Nix-specific
``LOCALE_ARCHIVE`` / ``NIX_LD`` / etc.) are preserved so the subprocess can
actually run.
"""

from __future__ import annotations

import os
from subprocess import Popen


def close_popen_pipes(proc: Popen) -> None:
    """Close a Popen's captured stdout/stderr pipes.

    E2E server fixtures capture the child's combined output on a
    ``stdout=PIPE`` pipe to surface its log on failure. The
    ``_io.BufferedReader`` that backs ``proc.stdout`` is *not* closed by
    ``proc.kill()/wait()``; leaving it open leaks an fd until GC, which
    surfaces as ``ResourceWarning: unclosed file`` in the suite's
    warnings summary (#1493). Call this at the end of every fixture's
    teardown after the log has been drained.
    """
    for stream in (proc.stdout, proc.stderr):
        if stream is not None:
            try:
                stream.close()
            except Exception:
                pass


# Prefixes whose vars are stripped: any env var starting with one of these
# (case-insensitive) is config or debug state that must not leak from the
# ambient env into a test subprocess.
_STRIP_PREFIXES = ("KLANGK", "_KLANGK", "KLANGKC", "LOGFIRE")

# Build-infra vars that locate *artifacts the test must use* (the workspace
# container image, the version stamp) — their values are produced by devenv's
# ``klangk:build-workspace-image`` task, not by any test, and every E2E server
# subprocess needs the real ones. These are forwarded from the ambient env
# deliberately (not stripped) so the server finds the built image.
# They are not test config — overriding one in a ``clean_env(...)`` call
# still wins.
#
# KLANGKD_FRONTEND_DIR is NOT here. Devenv no longer exports it (#1788; it's
# a klangkd.yaml setting), so each env-only launcher sets it explicitly
# (_e2e_server.start_server, global-setup.ts, run-demo-backend.sh). Forwarding
# a stray ambient value would risk pointing at a stale/wrong build.
_INFRA_VARS = (
    "KLANGKD_IMAGE_NAME",
    "KLANGKD_VERSION_FILE",
)


def clean_env(**overrides: str) -> dict[str, str]:
    """Return a hermetic env dict for a test subprocess.

    Starts from ``os.environ`` with every config-affecting var stripped, then
    applies ``overrides`` (the test's explicit KLANGKD_* / LOGFIRE_* keys).
    The baseline includes ``_KLANGKD_DISABLE_PROXY=1`` and
    ``KLANGKD_AUTH_MODES=password`` (the E2E default — most suites exercise the
    password auth flow); pass ``KLANGKD_AUTH_MODES="none"`` in overrides to
    opt into no-auth mode.

    Tests should call::

        env = clean_env(
            KLANGKD_PORT=port,
            KLANGKD_DATA_DIR=data_dir,
            ...
        )
    """
    env = {
        k: v
        for k, v in os.environ.items()
        if not k.upper().startswith(_STRIP_PREFIXES)
    }
    # Forward the build-infra vars (image / features / version stamp) so the
    # server subprocess finds the artifacts devenv built. See _INFRA_VARS.
    for name in _INFRA_VARS:
        val = os.environ.get(name)
        if val is not None:
            env[name] = val
    # If the caller pinned HOME, also pin XDG_CONFIG_HOME / XDG_STATE_HOME to
    # under it — unless the caller set them explicitly. Without this, the
    # inherited runner env (GitHub Actions ubuntu runners set
    # XDG_CONFIG_HOME=/home/runner/.config in /etc/environment; the value is
    # not re-expanded per-process) leaks past the HOME override, and any code
    # that correctly reads the XDG var (per spec) writes outside the tmpdir
    # the test set up. The CLI's config/state resolution (post-#1646) and the
    # server's (post-#1644) both honor these vars.
    if "HOME" in overrides and "XDG_CONFIG_HOME" not in overrides:
        env["XDG_CONFIG_HOME"] = f"{overrides['HOME']}/.config"
    if "HOME" in overrides and "XDG_STATE_HOME" not in overrides:
        env["XDG_STATE_HOME"] = f"{overrides['HOME']}/.local/state"
    # E2E baseline defaults.
    env["_KLANGKD_DISABLE_PROXY"] = "1"
    env.setdefault("KLANGKD_AUTH_MODES", "password")
    # #3157: per-client-IP API rate limiting is default-on in production
    # (300/60s). E2E suites generate machine-speed request bursts from one
    # loopback IP that would trip it nondeterministically; the limit itself
    # is exercised explicitly (test_api_rate_limit_e2e.py boots its own
    # server with a tiny budget).
    env.setdefault("KLANGKD_API_RATE_LIMIT", "0")
    # #3064: child CLI/TUI processes get the widened WS-connect wait on CI
    # to match the bring-up budgets (the strip above removes any ambient
    # value, so this stamp is the only way it reaches the child).
    if os.environ.get("CI"):
        env.setdefault("KLANGKC_WS_CONNECT_TIMEOUT", "240")
    env.update(overrides)
    return env


def ci_budget(default: float, ci: float) -> float:
    """Load-aware E2E budget, widened on CI (#3064).

    The four E2E suites share one runner VM; under that storage/IO
    contention a real podman bring-up (create/start/readiness) can
    outrun a tight local-dev budget — the observed failures were
    client-side 60s caps blowing while unrelated tests passed. Same
    shape as the frontend's container-ready doubling (#2745). Local
    runs keep the snappier default.
    """
    return ci if os.environ.get("CI") else default

File: src/frontend/e2e-tests/e2e-env.ts
/**
 * Hermetic env helper for E2E test suites (#1526).
 *
 * Mirrors `src/klangk/klangkd-tests/e2e-tests/_e2e_env.py`. Strips every config-affecting
 * prefix (KLANGK, _KLANGK, KLANGKC, LOGFIRE) from the ambient env so stray
 * vars can't leak into a test subprocess, then applies E2E baseline defaults
 * and the caller's overrides.
 */

const STRIP_PREFIXES = ["KLANGK", "_KLANGK", "KLANGKC", "LOGFIRE"];

// Build-infra vars that locate artifacts the test must use (the workspace
// container image, version stamp) — produced by devenv's
// klangk:build-workspace-image task, not by any test. Forwarded deliberately
// so the server finds the built image. (KLANGKBUILD_PLUGINS_DIR was removed
// in #1660/#1665 — the runtime reads features.json from the frontend bundle,
// not feature trees.)
//
// KLANGKD_FRONTEND_DIR is NOT here: devenv no longer exports it (#1788; it's
// a klangkd.yaml setting), so each env-only launcher sets it explicitly
// (global-setup.ts, run-demo-backend.sh). Forwarding a stray ambient value
// risks pointing at a stale/wrong build.
/**
 * The seeded admin password for E2E servers (KLANGKD_DEFAULT_PASSWORD).
 *
 * Must satisfy the boot-time password-policy gate (#2581): the daemon
 * refuses to start when the seeded admin's password violates
 * KLANGKD_MIN_PASSWORD_LENGTH (default 8) or any character-class count
 * (KLANGKD_PASSWORD_REQUIRE_*). "Admin123!" clears the length minimum and
 * all four classes, so it survives a policy tightening without a silent
 * 10-minute startup timeout in global-setup.
 */
export const ADMIN_PASSWORD = "Admin123!";

const INFRA_VARS = ["KLANGKD_IMAGE_NAME", "KLANGKD_VERSION_FILE"];

export function cleanEnv(
  overrides: Record<string, string> = {},
): Record<string, string> {
  const env: Record<string, string> = {};
  for (const [key, value] of Object.entries(process.env)) {
    if (value === undefined) continue;
    const upper = key.toUpperCase();
    if (STRIP_PREFIXES.some((p) => upper.startsWith(p))) continue;
    env[key] = value;
  }
  // Forward build-infra vars (image / version stamp / frontend).
  for (const name of INFRA_VARS) {
    const val = process.env[name];
    if (val !== undefined) env[name] = val;
  }
  // E2E baseline defaults. the proxy is NOT disabled: the frontend suite
  // launches real klangkd, fronted by its own proxy on KLANGKD_PORT (the
  // browser hits the proxy, which proxies to klangkd's UDS — #1525).
  if (!env.KLANGKD_AUTH_MODES) env.KLANGKD_AUTH_MODES = "password";
  Object.assign(env, overrides);
  return env;
}
